akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala

```diff
@@ -63,14 +64,34 @@ import scala.concurrent.duration.Deadline
     serviceRegistries(ddataKey).actorRefsFor(key)
   }
 
-  def activeActorRefsFor[T](key: ServiceKey[T], selfUniqueAddress: UniqueAddress): Set[ActorRef[T]] = {
+  /**
+   * @return keys that has a registered service instance on the given `address`
+   */
+  def keysFor(address: UniqueAddress)(implicit node: SelfUniqueAddress): Set[AbstractServiceKey] =
+    serviceRegistries.valuesIterator.flatMap(_.keysFor(address)).toSet
+
+  /**
+   * @return (reachable-nodes, all)
+   */
+  def activeActorRefsFor[T](
+      key: ServiceKey[T],
+      selfUniqueAddress: UniqueAddress): (Set[ActorRef[T]], Set[ActorRef[T]]) = {
     val ddataKey = ddataKeyFor(key)
     val entries = serviceRegistries(ddataKey).entriesFor(key)
     val selfAddress = selfUniqueAddress.address
-    entries.collect {
-      case entry if nodes.contains(entry.uniqueAddress(selfAddress)) && !hasTombstone(entry.ref) =>
-        entry.ref.asInstanceOf[ActorRef[key.Protocol]]
+    val (reachable, all) = entries.foldLeft((Set.newBuilder[ActorRef[T]], Set.newBuilder[ActorRef[T]])) {
+      case (builders @ (reachable, all), entry) =>
+        val entryAddress = entry.uniqueAddress(selfAddress)
+        if (nodes.contains(entryAddress) && !hasTombstone(entry.ref)) {
+          val ref = entry.ref.asInstanceOf[ActorRef[key.Protocol]]
+          all += ref
+          if (!unreachable.contains(entryAddress)) {
+            reachable += ref
+          }
+        }
+        builders
     }
+    (reachable.result(), all.result())
```

Comment:
Mixing mutability and FP like there's no tomorrw. Sorry!

Reply:
maybe better to go all in, using only mutable in a `foreach` instead of `foldLeft`?